doc/cuebreakpoints.1
.TH "cuebreakpoints" "1"
.SH NAME
cuebreakpoints \- report track breakpoints from a CUE or TOC file
.SH SYNOPSIS
.B cuebreakpoints
[ {
.B \-i
.I format
|
.BR \-\-input\-format =\fIformat\fP
} {
.B \-\-append\-gaps
|
.B \-\-prepend\-gaps
|
.B \-\-split\-gaps
} ]
[
.I file
\&... ]
.br
.B cuebreakpoints \-h | \-\-help
.br
.B cuebreakpoints \-V | \-\-version
.SH DESCRIPTION
.B cuebreakpoints
reports the track breakpoints found in CUE and TOC files, which are
typically created by compact disc ripping software.
The breakpoints are in a format usable by
.B shnsplit
(part of the shntool package).
.PP
Three approaches to track pregaps are available: append (to previous track),
prepend (to succeeding track), and split (from both preceding and
succeeding tracks).
The split approach can result in up to twice as many breakpoints being
reported as there are tracks on the disc.
.PP
The first track's pregap cannot be appended to the previous track, so it is
prefixed to the track in both append and prepend modes.
If you want the track without it, use the
.B \-\-split\-gaps
option.
.PP
If no filenames are specified,
.B cuebreakpoints
reads from standard input, and an input format option
.I must
be specified.
If one or more filenames is provided, but the input format option is not
specified, the input format will be guessed based on each file's suffix
(e.g.,
.I .cue
or
.IR .toc ).
This heuristic is case-insensitive.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
displays a usage message and exits.
.TP
.BR \-i " \fIformat\fP, " \-\-input\-format=\fIformat\fP
sets the expected format of the input file(s) to
.IR format ,
which must be either
.B cue
or
.BR toc .
.TP
.B \-\-append\-gaps
appends pregaps to the end of the previous track.
This is the default.
.TP
.B \-\-prepend\-gaps
prefixes pregaps to the beginning of each subsequent track.
.TP
.B \-\-split\-gaps
separates pregaps from both the preceding and succeeding tracks.
.TP
.B \-V, \-\-version
displays version information and exits.
.PP
If more than one of
.BR \-\-append\-gaps ,
.BR \-\-prepend\-gaps ,
and
.B \-\-split\-gaps
are specified, all except the last encountered are ignored.
.SH "EXIT STATUS"
.B cuebreakpoints
exits with status zero if it successfully generates a report for each
input file, and nonzero if there were problems.
.SH AUTHOR
Cuetools was written by Svend Sorensen.
Branden Robinson contributed fixes and enhancements to the utilities and
documentation.
.SH "SEE ALSO"
.BR cueconvert (1),
.BR cueprint (1)


NEWS
Changes since 1.3

* Applied most of Branden Robinson's Debian patch
  (cuetools_1.3-2.diff.gz). This included many fixes and enhancements
  to tools and manual pages. cuebreakpoints and cueconvert will exit if
  they fail on any of the input files. cuebreakpoints, cueconvert, and
  cuetools will exit with a nonzero return value if there was a failure
  on an input file.

* Added version option to all tools.

* Single quoted string support has been removed.  This was broken, and I am not
  sure of any software that supports single quoted strings in cue or toc files.

Changes since 1.2

* Build system has been rewritten using automake/autoconf.

* Added append, prepend, and split pregap modes to cuebreakpoints.

* Fixed possible compilation error with Bison.

* All tools accept long options.

* Programs exit if --input-format or --output-format is an illegal value.

Changes since 1.1

* cuebreakpoints was not printing the last track breakpoint.  This has been
  fixed.

* cueprint now prints a zero-length string for undefined values.  Flags, width,
  and precision are applied to conversion characters which have no conversion.

* Added a tagging script, cuetag.sh, which uses cueprint, vorbiscomment, and
  mp3info to tag Ogg Vorbis and MP3 files.  This needs more testing and
  documentation.

* Added a track selection flag to cueprint.

Changes since 1.0

* File formats documentation has been readded and updated (docs/formats.txt).

* ``CDTEXTFILE`` is now a recognized (but ignored) keyword.

* If only a disc or track template is specified on the command line, then only
  the disc or track information is printed, respectively.

* cueprint now supports flags and width and precision values for template
  expansion (like the printf command).

Changes since 0.6

* cuetools-1.x is a significant code rewrite of 0.x.  The cue/toc parsing code
  has been reimplemented in lex/yacc.

* cuerename has been replaced by the more general utility cueprint.

* Usage syntax of the utilities has changed.  See the respective man pages.


src/tools/cuebreakpoints.c
/*
 * cuebreakpoints.c -- print track break points
 *
 * Copyright (C) 2004, 2005, 2006, 2007 Svend Sorensen
 * For license terms, see the file COPYING in this distribution.
 */

#include <getopt.h>	/* getopt_long() */
#include <stdio.h>	/* fprintf(), printf(), snprintf(), stderr */
#include <stdlib.h>	/* exit() */
#include <string.h>	/* strcasecmp() */
#include "cuefile.h"
#include "time.h"

#if HAVE_CONFIG_H
#include "config.h"
#else /* not HAVE_CONFIG_H */
#define PACKAGE_STRING "cuebreakpoints"
#endif /* HAVE_CONFIG_H */

char *progname;

/*
 * pregap correction modes:
 * APPEND - append pregap to previous track (except for first track)
 * PREPEND - prefix pregap to current track
 * SPLIT - print breakpoints for beginning and end of pregap
 */
enum GapMode {APPEND, PREPEND, SPLIT};

/* Print usage information and exit */
void usage(int status)
{
	if (0 == status) {
		printf("Usage: %s [option...] [file...]\n", progname);
		printf("Report track breakpoints from a CUE or TOC file.\n"
		       "\n"
		       "OPTIONS\n"
		       "-h, --help			print usage\n"
		       "-i, --input-format cue|toc	set format of file(s)\n"
		       "--append-gaps			append pregaps to previous track (default)\n"
		       "--prepend-gaps			prefix pregaps to track\n"
		       "--split-gaps			split at beginning and end of pregaps\n"
		       "-V, --version			print version information\n");
	} else {
		fprintf(stderr, "Try `%s --help' for more information.\n", progname);
	}

	exit (status);
}

/* Print version information and exit */
void version()
{
	printf("%s\n", PACKAGE_STRING);

	exit(0);
}

void print_m_ss_ff(long frame)
{
	int m, s, f;

	time_frame_to_msf(frame, &m, &s, &f);
	printf ("%d:%02d.%02d\n", m, s, f);
}

void print_breakpoint(long b)
{
	/* Do not print zero breakpoints. */
	if (0 != b) {
		print_m_ss_ff(b);
	}
}

void print_breaks(Cd *cd, int gaps)
{
	int i;
	long b;
	long pg;
	Track *track;

	for (i = 1; i <= cd_get_ntrack(cd); i++) {
		track = cd_get_track(cd, i);
		/*
		 * When breakpoint is at:
		 * index 0: gap is prepended to track
		 * index 1: gap is appended to previous track
		 */
		b = track_get_start(track);
		pg = track_get_index(track, 1) - track_get_zero_pre(track);

		if (gaps == PREPEND || gaps == SPLIT) {
			print_breakpoint(b);
		/*
		 * There is no previous track to append the first track's
		 * pregap to.
		 */
		} else if (gaps == APPEND && 1 < i) {
			print_breakpoint(b + pg);
		}

		/* If pregap exists, print breakpoints (in split mode). */
		if (gaps == SPLIT && 0 < pg) {
			print_breakpoint(b + pg);
		}
	}
}

int breaks(char *name, int format, int gaps)
{
	Cd *cd = NULL;

	if (NULL == (cd = cf_parse(name, &format))) {
		fprintf(stderr, "%s: error: unable to parse input file"
		        " `%s'\n", progname, name);
		return -1;
	}

	print_breaks(cd, gaps);

	return 0;
}

int main(int argc, char *argv[])
{
	int format = UNKNOWN;
	int gaps = APPEND;
	int ret = 0;		/* return value of breaks() */

	/* option variables */
	int c;
	/* getopt_long() variables */
	extern char *optarg;
	extern int optind;

	static struct option longopts[] = {
		{"help", no_argument, NULL, 'h'},
		{"input-format", required_argument, NULL, 'i'},
		{"append-gaps", no_argument, NULL, 'a'},
		{"prepend-gaps", no_argument, NULL, 'p'},
		{"split-gaps", no_argument, NULL, 's'},
		{"version", no_argument, NULL, 'V'},
		{NULL, 0, NULL, 0}
	};

	progname = argv[0];

	while (-1 != (c = getopt_long(argc, argv, "hi:V", longopts, NULL))) {
		switch (c) {
		case 'h':
			usage(0);
			break;
		case 'i':
			if (0 == strcmp("cue", optarg)) {
				format = CUE;
			} else if (0 == strcmp("toc", optarg)) {
				format = TOC;
			} else {
				fprintf(stderr, "%s: error: unknown input file"
				        " format `%s'\n", progname, optarg);
				usage(1);
			}
			break;
		case 'a':
			gaps = APPEND;
			break;
		case 'p':
			gaps = PREPEND;
			break;
		case 's':
			gaps = SPLIT;
			break;
		case 'V':
			version();
			break;
		default:
			usage(1);
			break;
		}
	}

	/* What we do depends on the number of operands. */
	if (optind == argc) {
		/* No operands: report breakpoints of stdin. */
		ret = breaks("-", format, gaps);
	} else {
		/* Report track breakpoints for each operand. */
		for (; optind < argc; optind++) {
			ret = breaks(argv[optind], format, gaps);
			/* Exit if breaks() returns nonzero. */
			if (!ret) {
				break;
			}
		}
	}

	return ret;
}
